media_driver/agnostic/common/os/mos_sw_swizzle.c

```diff
@@ -0,0 +1,1029 @@
+* copy of this software and associated documentation files (the "Software"),
+* to deal in the Software without restriction, including without limitation
+* the rights to use, copy, modify, merge, publish, distribute, sublicense,
+* and/or sell copies of the Software, and to permit persons to whom the
+* Software is furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included
+* in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+* OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+* THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+* OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+* ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+* OTHER DEALINGS IN THE SOFTWARE.
+*/
+//!
+//! \file     mos_sw_swizzle.c
+//! \brief    Mesa ISL (Intel Surface Layout) implementation for sw-based tiling/detiling
+//! \details  Mesa ISL (Intel Surface Layout) implementation for sw-based tiling/detiling. SSE optimized
+//!
+
+#include <string.h>
+#include <immintrin.h>
+#include "mos_sw_swizzle.h"
+
+#define INLINE_SSE41
+#define USE_SSE41
+
+typedef void *(*isl_mem_copy_fn)(void *dest, const void *src, size_t n);
+
+static inline int
+_mesa_is_pow_two(int x)
+{
+   return !(x & (x - 1));
+}
+
+
+static inline uintptr_t
+ALIGN(uintptr_t value, int32_t alignment)
+{
+   MOS_OS_ASSERT((alignment > 0) && _mesa_is_pow_two(alignment));
+   return (((value) + (alignment) - 1) & ~((alignment) - 1));
+}
+
+static inline uintptr_t
+ROUND_DOWN_TO(uintptr_t value, int32_t alignment)
+{
+   MOS_OS_ASSERT((alignment > 0) && _mesa_is_pow_two(alignment));
+   return ((value) & ~(alignment - 1));
+}
+
+/**
+ * Unreachable macro. Useful for suppressing "control reaches end of non-void
+ * function" warnings.
+ */
+#if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5)
+#define unreachable(str)    \
+do {                        \
+   MOS_OS_ASSERT(!str);            \
+   __builtin_unreachable(); \
+} while (0)
+#elif defined (_MSC_VER)
+#define unreachable(str)    \
+do {                        \
+   MOS_OS_ASSERT(!str);            \
+   __assume(0);             \
+} while (0)
+#else
+#define unreachable(str) MOS_OS_ASSERT(!str)
+#endif
+
+#ifdef __GNUC__
+#define FLATTEN __attribute__((__flatten__))
+#else
+#define FLATTEN
+#endif
+
+/* Forced function inlining */
+/* Note: Clang also sets __GNUC__ (see other cases below) */
+#ifndef ALWAYS_INLINE
+#  if defined(__GNUC__)
+#    define ALWAYS_INLINE inline __attribute__((always_inline))
+#  elif defined(_MSC_VER)
+#    define ALWAYS_INLINE __forceinline
+#  else
+#    define ALWAYS_INLINE inline
+#  endif
+#endif
+
+/** Minimum of two values: */
+#define MIN2( A, B )   ( (A)<(B) ? (A) : (B) )
```

Comment:
media driver is mostly written in C++, but this file is C and you use some MOS_ASSERT stuff below. We potentially may step into problems if these macro defs will start using C++. Please, align with code maintainers whether you should use C or C++. 

If you will decide to use C++, consider use std::min/max and other primitives.

Reply:
@XinfengZhang , please say what is preferable

Reply:
c++ is preferable